src/domain/usecases/CopyDataSetsOUUserCase.ts

```diff
@@ -0,0 +1,96 @@
+import _ from "lodash";
+import { DataSetsRepository } from "domain/repositories/DataSetsRepository";
+import { PostOptions, Ref } from "types/d2-api";
+import { DataSet } from "domain/entities/DataSet";
+import { OUCopyResult } from "domain/entities/OUCopyResult";
+
+export class CopyDataSetsOUUserCase {
+    constructor(
+        private dataSetsRepository: DataSetsRepository
+    ) { }
+
+    async execute(options: {
+        originDataset: string;
+        destinationDatasets: string[];
+        replace: boolean | undefined;
+    }): Promise<OUCopyResult> {
+        const { originDataset, destinationDatasets, replace } = options;
+
+        console.debug(`Replace the destination OU: ${replace ? "True" : "False"}`);
+
+        const datasets = await this.dataSetsRepository.get([originDataset, ...destinationDatasets]);
+
+        const origDataset = datasets[originDataset];
+        const destDatasets = destinationDatasets.map((id) => datasets[id]);
+
+        if (!origDataset || !destDatasets) throw new Error("Missing DataSets");
+
+        const data: DataSet[] = [];
+
+        destDatasets.forEach((destDataSet) => {
+            if (!destDataSet) throw new Error("Missing DataSets");
+            const dataSetsEqual = compare(origDataset, destDataSet, replace ?? false);
+
+            if (!dataSetsEqual) {
+                if (replace) {
+                    data.push({ ...destDataSet, organisationUnits: origDataset.organisationUnits })
+                } else {
+                    data.push(mergeDataSetOUs(origDataset, destDataSet));
+                }
+            } else {
+                console.debug(`DataSet with ID:${destDataSet.id} already contains all the OUs.`);
+            }
+        });
+
+        let result: OUCopyResult;
+        if (!_.isEmpty(data)) {
+            const metadata = { dataSets: [...data] };
```

Comment:
Why the copy? `[...data]`. It should not be touched by the post.

Reply:
Indeed, no need for the shallow copy.